### DISCUSSION TITLE
How to set up Multi vendor marketplace - 1 product many vendors

### DISCUSSION THREAD

**Discussion by onwosystems** [swellstores/community#92] [Help] [answered]
Hi All,

We are looking to set up a multi vendor marketplace.
1. A Product has options. => We generate variants.  Each product:variant  may be sold by multiple vendors. 
2. Each vendor may have his own pricing for a given product:variant

The Swell Marketplace guide showcases how to set up a marketplace where a Product is sold by 1 vendor alone. 

We could use this model too, where we create multiple Products instead of 1.  
Meaning, if "Red T Shirt" is sold by Vendor1 and Vendor2, we will have to create 2 products. "Red T Shirt" by vendor1 and "Red T shirt" by vendor 2. We will then need a way to figure out how to display a single Product in Browse page etc. Doable - but sounds like a hack.

Instead, we would like to extend the standard Swell model and introduce vendor and their pricing as a concept.  One advantage of doing this is that we should be able to fetch the "possible vendors and their prices" as a single product API call. 

One catch we see is that, the standard add-to-cart flows etc automatically work by using the price in the product:variant. Since 1 product:variant is sold by multiple vendors, we obviously will have to keep that as an object within it, or a linked model. And during add-to-cart, we will have to invoke a back-end call to correctly figure out which vendor was selected and therefore what price should be applied.  Again, sounds a bit hackish. 

Would anyone have any suggestions on how to do this cleanly?

**Comment by sags95** [comment, marked as answer]
@onwosystems  You should be able to accomplish this using product variants and by creating a `vendor_id` field on product variants through the Model Editor. These variants can be created using the same `product_options_ids` as well, but they'll be unique as they will have different variant_id, vendor_ids, prices, inventory, etc. In this case each variant is managed by a different vendor.
<img width="1458" alt="Screenshot 2023-01-18 at 4 21 37 PM" src="https://user-images.githubusercontent.com/9522171/213298172-dda25a09-1329-476e-b9e0-78f180ffb7e0.png">

<img width="1474" alt="Screenshot 2023-01-18 at 4 25 33 PM" src="https://user-images.githubusercontent.com/9522171/213298327-333ab6d5-81a3-4a13-bb4f-9fc7cb3666a6.png">

**Reply by onwosystems** [reply]
Thank you for this.

If I understood correctly, 
A: add a vendor_id on variants in model.  However, this means a given variant can be sold by only 1 vendor (in practice). Because each Option combination has a vendor field. 
B: Add vendor as an Option. This way, Swell auto generates combinations for each vendor. Advantage is that the price, inventory can be set correctly as business situation demands.  Glitch is that this can explode variant count. For egs: if there are 3 options with 2x3x4 values each, and say 5 vendors selling these, that means 2x3x4x5 variants.

**Reply by sags95** [reply]
@onwosystems If you would like each variant to be sold by multiple vendors, you can also create a collection of vendors (accounts) instead of just 1 vendor field. However, if the vendors are managing the prices and details on their side, typically we would have 1 variant relate to 1 vendor. In my example above, those 2 variants have the same options (so similar variants), however they would be linked to 2 vendors that can manage the prices and inventory.

**Reply by logeshswell** [reply]
@onwosystems cc: @sags95 I see `vendor_id` field in variants model seems redundant.
Adding vendor as an option looks very good to me as it offers enough flexibility to query the variants corresponding to a vendor along with pricing and inventory management. 

> Glitch is that this can explode variant count. For egs: if there are 3 options with 2x3x4 values each, and say 5 vendors selling these, that means 2x3x4x5 variants.

Also, I feel having a larger variant count shouldn't be a big deal, as there's a chance of all 5 vendors have a completely different pricing. In that case, nx5 variants with unique pricing need to be stored in our system.
